### DISCUSSION TITLE
Trace Id from manual logs into Seq

### DISCUSSION THREAD

**Discussion by BrettGoreham** [datalust/seq-tickets#2124] [Help] [answered]
Hello I am trying to send logs to sec that are being sent in from manual logging from pino is javascript.

We are struggling to get trace id to be present in these logs. We have tried to to pass the trace id as a property but no matter what we try (@TraceId, TraceId, trace-id) we can not manage to get Seq to pick it up as the trace id.

This results in the query @TraceId = 'traceidString'  to not pick up these logs.
for example this query will get both when sending @TraceId as a log property
@TraceId = 'PicksUpBackendLogs' or @Properties['@TraceId'] = 'PicksUpFrontEndLogs'.

Is there a magic way to get the front end logs to be picked up by the built in @TraceId?

Thanks for any help!

**Comment by liammclennan** [comment]
Hi Brett,
None of the JavaScript clients support tracing, except https://github.com/open-telemetry/opentelemetry-js

**Reply by BrettGoreham** [reply]
@liammclennan Thats great but there is no manual way to add trace Id to logs created manually?

That would be a handy feature to have if other libraries decide not to support tracing, as I doubt we will change libraries just for tracing.

Thanks for the help anyway!

**Reply by liammclennan** [reply]
You can add a correlation id (like traceid), and use it to group activities within a transaction. Activating the tracing features in Seq requires more than just a trace id, and it requires support within the client library.

**Comment by nblumhardt** [comment]
Hi @BrettGoreham - quick question, what are you using to generate the trace ids, and what's the scope of a trace/span in your client code?

We're putting our heads together trying to come up with some pleasant ways to enable this, but need to understand the landscape a bit better...

Thanks!

**Reply by BrettGoreham** [reply]
Hi we are using open telemetry to create spans/traces but struggle to send these along in the logs we are logging with pino.

I tried https://www.npmjs.com/package/@opentelemetry/instrumentation-pino 
however this didnt seem to have effects on the logs that were being sent.

But I must note im a backend developer so there is a strong possibility I am understanding something wrong

**Reply by nblumhardt** [reply]
Thanks for the link; you're not missing anything, it's an evolving area so we're still piecing a lot of the picture together ourselves :+1: 

I think what's needed to make this happen is:

1. The `datalust/seq-logging` package needs to be extended to accept `traceId` and `spanId` properties on the logged event objects, and [map these to `TraceId` and `SpanId` properties on the API payload](https://github.com/datalust/seq-logging/blob/dev/seq_logger.js#L118)
2. The `datalust/pino-seq` package needs to pass `traceId` and `spanId` properties on the `forSeq` object [here](https://github.com/datalust/pino-seq/blob/dev/pinoSeqStream.js#L43) - these can come from `props.trace_id` and `props.span_id`, to match what the OpenTelemetry plug-in is adding by default.

Should be pretty straightforward, I hope; we'll have a chat here about when we might be able to tackle this, but if you're interested in sending a PR that would also be welcome, just let us know if so.

Thanks again for all the info!

**Comment by liammclennan** [comment]
Hi @BrettGoreham 

How are you sending the logs to Seq? Are you using `pino-seq` or the OpenTelemetry OTLP exporter?

**Reply by BrettGoreham** [reply]
Hi Again we are using pino-seq for logs for the most part,  I tried briefly with the exporter after reading through documentation but didnt get results after a quick test

**Reply by liammclennan** [reply]
Hi @BrettGoreham 

Thanks for the update. I'm working on a [PR](https://github.com/datalust/pino-seq/pull/34) to get trace id and span id to come through `pino-seq`. Note that this will not be enough to make the Seq trace viewer work. I might also see if I can get Pino to seq logs to Seq via the OTLP exporter.

**Reply by liammclennan** [reply, marked as answer]
Hi @BrettGoreham 

This is now published as `pino-seq` 1.1.0.

If your log events have the properties `trace_id` and `span_id` this will be available in Seq as `@TraceId` and `@SpanId` respectively.

**Reply by antoniaelek** [reply]
Hi, I am trying to include trace id in log eventss using .NET but seems to me it is just not possible. I have tried setting `@Trace` property and also `Trace`, and all kinds of escaping, all to no avail. Is this supported? Should it be?

**Reply by liammclennan** [reply]
Hi @antoniaelek 

You will need a client that is capable of tracing. For .NET that is either the .NET OpenTelemetry SDK, or SerilogTracing.

Here is an example with the OpenTelemetry SDK: https://github.com/datalust/seq-examples/tree/main/client/opentelemetry-csharp-sdk-to-seq

Here is an example with SerilogTracing: https://github.com/serilog-tracing/serilog-tracing/tree/dev/example

There is no way to manually add a trace id to a log event and have it recognized as being part of a trace.

**Reply by antoniaelek** [reply]
Hi @liammclennan 
I am using OpenTelemetry SDK for sending traces to seq and this works fine. I was just thinking since it looks like it is possible in JS to have `@TraceId` in log messages also, it could be possible in .net too.
You can see in the screenshot that traces have TraceId set but i would like to have it also on log messages, to be able to get as much information as possible when querying by it.
![image](https://github.com/user-attachments/assets/f9e036c3-53c2-4a7f-a333-6dae90bd20b5)

**Reply by liammclennan** [reply]
Logs that are generated inside a span do get that span and trace id. Span Id and Trace Id are not shown as properties by Seq, but you can view them as columns. Also, if you choose 'Find' for a trace or span the logs will be in the search results.

**Reply by antoniaelek** [reply]
I see. However, I am unable to achieve this behavior. Am I doing something wrong? My code looks like this (log statement is inside span) but still does not have TraceId column, as seen by missing green column in log event (2nd row) in the screenshot.
```csharp
using (var activity = ActivitySource.StartActivity($"RabbitMQ producer publish {@event.Type} message {@event.Id}", ActivityKind.Producer))
{
    // ...
    _logger.LogInformation(
        "Published event {EventId} to RabbitMQ exchange {ExchangeName} using routing key {EventName}.",
        @event.Id, ExchangeName, @event.Type);
}
```
![image](https://github.com/user-attachments/assets/9b80166e-0d1b-4107-9698-8dfe2404f7da)

**Reply by liammclennan** [reply]
@antoniaelek you can check which events have @TraceId with a query like: `select @TraceId, @SpanId, @Message from stream where @Start is null limit 1000`. 

If you confirm that @TraceId and @SpanId are not there then I suggest reviewing https://github.com/datalust/seq-examples/blob/main/client/opentelemetry-csharp-sdk-to-seq/Example.WeatherService/Program.cs (which works) and see how it differs from your usage.

**Reply by antoniaelek** [reply]
It works when configuring logging without Serilog (as in example you provided), but when using Serilog as logging engine, and configuring Serilog.Sinks.OpenTelemetry, it still does not add trace id and span id to log events, just traces.

**Reply by liammclennan** [reply]
Serilog logging won't pickup trace and span information from open telemetry because they are different libraries. You should use the builtin logging - or remove Open Telemetry entirely and replace it with SerilogTracing.

**Reply by antoniaelek** [reply]
Okay, will do that, then. Thanks for your help.

**Reply by liammclennan** [reply]
You're welcome.
